File: ShapeRain/src/Star.java
import processing.core.PApplet;
import processing.core.PVector;

public class Star extends Polygon {

    public Star(PApplet p, PVector position, int points, float radius) {
        super(p, position, points, radius);
    }

    protected void initVertices(int points) {
        float theta = 0;
        points = PApplet.constrain(points, 3, 20);
        vertices = new PVector[points * 2];
        float deltaTheta = p.TWO_PI / points * 2;

        for (int i = 0; i < vertices.length; i++) {
            float r = radius;

            if (i % 2 == 0) {
                r /= 2;
            }

            float x = PApplet.cos(theta) * r;
            float y = PApplet.sin(theta) * r;
            theta += deltaTheta;
            vertices[i] = new PVector(x, y);
        }
    }
}

File: docs/ShapeRain/shape/star.js
class Star extends Polygon {

    constructor(postition, radius, pointCount) {
        super(postition, radius, pointCount);
    }

    _initVertices(pointCount) {
        let theta = 0;
        pointCount = constrain(pointCount, 3, 20);
        let deltaTheta = TWO_PI / (pointCount * 2);

        for (let i = 0; i < pointCount * 2; i++) {
            let r = this.getRadiusForIndex(i);
            let x = cos(theta) * r;
            let y = sin(theta) * r;
            theta += deltaTheta;
            this.vertices.push(createVector(x, y));
        }
    }

    getRadiusForIndex(index) {
        let r = this.radius;

        if (this.isEven(index)) {
            r /= 2;
        }

        return r;
    }

    isEven(index) {
        return index % 2 === 0;
    }
}
